File: cmd_init.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strings"

	"github.com/nektro/go-util/util"
	"github.com/spf13/cobra"
)

func init() {
	last := func(a []string) string {
		if len(a) == 0 {
			return ""
		}
		return a[len(a)-1]
	}
	tryIndex := func(a []string, n int) string {
		if n >= len(a) {
			return ""
		}
		return a[n]
	}
	detectPkgName := func(arg string) string {
		if len(arg) > 0 {
			return arg
		}
		n, _ := filepath.Abs("./")
		n = last(strings.Split(n, "/"))
		n = strings.TrimPrefix(n, "zig-")
		assert(len(n) > 0, "name may not be an empty string")
		return n
	}
	detectMainFile := func(arg string) string {
		if len(arg) > 0 {
			argR, err := filepath.Abs(arg)
			util.DieOnError(err)
			if !util.DoesFileExist(argR) {
				util.DieOnError(errors.New("specified entry point file does not exist: " + arg))
			}
			if !strings.HasSuffix(arg, ".zig") {
				util.DieOnError(errors.New("main entry point must be a .zig file"))
			}
			cwd, _ := os.Getwd()
			argR = strings.TrimPrefix(argR, cwd)[1:]
			argR = strings.ReplaceAll(argR, "\\", "/")
			return argR
		}
		if util.DoesFileExist("./src/main.zig") {
			return "src/main.zig"
		}
		util.DieOnError(errors.New("unable to determine package entry point"))
		return ""
	}
	rootCmd.AddCommand(&cobra.Command{
		Use:   "init",
		Short: "Initialize a new package.",
		Long:  `zigmod init [name] [main_file]`,
		Run: func(cmd *cobra.Command, args []string) {
			buf := new(bytes.Buffer)

			name := detectPkgName(tryIndex(args, 0))
			fmt.Fprintln(buf, "name:", name)

			mainf := detectMainFile(tryIndex(args, 1))
			fmt.Fprintln(buf, "main:", mainf)

			fmt.Fprintln(buf, "dependencies:")
			f, err := os.Create("./zig.mod")
			util.DieOnError(err)

			fmt.Fprint(f, buf.String())
			log.Println("Initialized a new package named", name, "with entry point", mainf)
		},
	})
}

File: src/cmd_init.zig
const std = @import("std");
const gpa = std.heap.c_allocator;

const u = @import("./util/index.zig");

//
//

pub fn execute(args: [][]u8) !void {
    const name = try detect_pkgname(u.try_index([]const u8, args, 0, ""));
    const mainf = try detct_mainfile(u.try_index([]const u8, args, 1, ""));

    const file = try std.fs.cwd().createFile("zig.mod", .{});
    defer file.close();

    const fwriter = file.writer();
    try fwriter.print("id: {s}\n", .{u.random_string(48)});
    try fwriter.print("name: {s}\n", .{name});
    try fwriter.print("main: {s}\n", .{mainf});
    try fwriter.print("dependencies:\n", .{});

    u.print("Initialized a new package named {s} with entry point {s}", .{name, mainf});
}

fn detect_pkgname(def: []const u8) ![]const u8 {
    if (def.len > 0) {
        return def;
    }
    const dpath = try std.fs.cwd().realpathAlloc(gpa, "build.zig");
    const split = try u.split(dpath, std.fs.path.sep_str);
    var name = split[split.len-2];
    name = u.trim_prefix(name, "zig-");
    u.assert(name.len > 0, "package name must not be an empty string", .{});
    return name;
}

fn detct_mainfile(def: []const u8) ![]const u8 {
    if (def.len > 0) {
        if (try u.does_file_exist(def)) {
            if (std.mem.endsWith(u8, def, ".zig")) {
                return def;
            }
        }
    }
    if (try u.does_file_exist(try std.fs.path.join(gpa, &[_][]const u8{"src", "lib.zig"}))) {
        return "src/lib.zig";
    }
    if (try u.does_file_exist(try std.fs.path.join(gpa, &[_][]const u8{"src", "main.zig"}))) {
        return "src/main.zig";
    }
    u.assert(false, "unable to detect package entry point", .{});
    unreachable;
}
